cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkSubmitter.java

```diff
@@ -0,0 +1,180 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.spark;
+
+import co.cask.cdap.api.spark.SparkSpecification;
+import co.cask.cdap.common.lang.ClassLoaders;
+import co.cask.cdap.common.lang.WeakReferenceDelegatorClassLoader;
+import co.cask.cdap.internal.app.runtime.distributed.LocalizeResource;
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.Uninterruptibles;
+import org.apache.spark.deploy.SparkSubmit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * Provides common implementation for different {@link SparkSubmitter}.
+ */
+public abstract class AbstractSparkSubmitter implements SparkSubmitter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractSparkSubmitter.class);
+
+  // Filter for getting archive resources only
+  private static final Predicate<LocalizeResource> ARCHIVE_FILTER = new Predicate<LocalizeResource>() {
+    @Override
+    public boolean apply(LocalizeResource input) {
+      return input.isArchive();
+    }
+  };
+
+  // Transforms LocalizeResource to URI string
+  private static final Function<LocalizeResource, String> RESOURCE_TO_PATH = new Function<LocalizeResource, String>() {
+    @Override
+    public String apply(LocalizeResource input) {
+      return input.getURI().toString();
+    }
+  };
+
+  @Override
+  public final <V> ExecutionFuture<V> submit(final ExecutionSparkContext sparkContext, Map<String, String> configs,
+                                             List<LocalizeResource> resources, File jobJar, final V result) {
+
+    final SparkSpecification spec = sparkContext.getSpecification();
+    final List<String> args = createSubmitArguments(spec.getMainClassName(), configs, resources, jobJar);
+
+    // Spark submit is called from this executor
+    // Use an executor to simplify logic that is needed to interrupt the running thread on stopping
+    final ExecutorService executor = Executors.newSingleThreadExecutor(new ThreadFactory() {
```

Comment:
Can't the executor thread ties to the submit call. The same submitter can have multiple spark job submitted through it (e.g. in Workflow fork).